Added Breadcrumbs

hokei opened the pull request:
[ResearchNow Jira Ticket](http://jira.nypl.org/browse/SFR-694)

## **This PR does the following:**
- Adds the React version of the breadcrumb

### Front End Review:
- [ ] View [the example in Storybook](https://reno-XXX-nypl.pantheonsite.io/themes/custom/nypl_emulsify/pattern-lab/public)
- [ ] Check against the [Metronome documentation](http://themetronome.co/components/mn-11-breadcrumb-navigation/?fresh=true)

hokei committed:
Added Breadcrumbs

hokei commented:
In addition to wanting to know your thoughts about the approach, I'm also wondering where the breadcrumb background comes from?  Both the Twig and React versions of this have white text, and in patternlab/storybook this shows up as white text on a white background, which is not helpful.  

There's also an icon at the end, which doesn't show up in the Metronome to me 

This is not finished, but I think before working further I'd love to clarify what we want from the storybook re:knobs, variations, previews, etc!

helenvholmes commented:
We chatted about where the background-color comes from in person yesterday, but I'm going to capture that info here.

The background-color is determined by the section that contains the component. For example, in Research, all of the components that have section-based colors—breadcrumbs, article card-no image, etc.—will update to be the Research teal. In the abstract, this means the breadcrumbs _have_ no background-color. In Patternlab we handle this by having Patternlab-specific styles. I think the in the future it means that we'll probably mock up data that includes the section it's in.

Since none of that work has been completed, it means for the time being we should probably have some Storybook-specific CSS as well that's only added and used in the stories. You can see an example of that [here](https://github.com/NYPL/nypl-d8/blob/e752f642775d381aa58010b1aceebf175fa6cf7b/web/themes/custom/nypl_emulsify/components/_patterns/02-molecules/menus/breadcrumbs/_breadcrumbs.scss#L68).

helenvholmes commented on src/react-components/stories/1-Button.stories.tsx:
```diff
@@ -1,6 +1,6 @@
 import * as React from "react";
 
-import Button from "../components/Button/Button";
+import Button from "../src/components/Button/Button";
```
Not a line comment but a filename comment. Why are we prefixing the story file with a "1"? I was under the assumption that this was how you ordered stories in the file viewer. If we're fine with letting them fall alphabetically, we can let them be unprefixed.

If we'd like them to begin matching the structure in the Twig storybook, we could probably begin mimicking the file structure in the other places of `atoms/`, `molecules`, and `organisms`, and within those folders not numbering them.

hokei replied:
I think it would be better to mimic the Twig structure since it's the same structure as in Metronome, right?  Also could take this time to move stories to be in the same folder as the tsx.

helenvholmes replied:
I think it's best to stay consistent across our three packages. It's a lot of folders, but at least you always know where you're going instead of `find . --name`ing for files.

helenvholmes commented on src/react-components/stories/1-Breadcrumb.stories.tsx:
```diff
@@ -0,0 +1,11 @@
+  title: "Breadcrumb",
+};
+
+export const regular = () => <Breadcrumb items={[{ url:'#', text:"Home"}, { url:'#', text:"Parent"}]}></Breadcrumb>;
```
Can you added two more levels to show off what happens when there are more than two items? "Grandparent" and "Great-Grandparent"?

Might be more readable to have the key/value pairings as a blog with a variable name outside of the function so adding on to it is easier in the future if we want to generate some stories for this that _do_ include a ton of links.

helenvholmes commented on src/react-components/src/components/Breadcrumb/Breadcrumb.tsx:
```diff
@@ -0,0 +1,42 @@
+render(): JSX.Element {
+    const { items, className } = this.props;
+
+    let bcrumbClassName = `breadcrumb${className ? `${className}` : ""}`
```
If we need to pass multiple classes to a breadcrumb in a story (say, we need to pass in a series of modifiers like `--section--books-and-more`, how does that work? Relevant to [the background discussion](https://github.com/NYPL/nypl-design-system/pull/40#issuecomment-550347049).

hokei replied:
AFAICT you can't do that with the Twig right now either, and I think for now it's ok to leave it at parity.

helenvholmes replied:
So you can do it with Twig right now—it's what the `bem` function does. I was curious how that worked here

hokei replied:
So I think I might be misunderstanding how to do this 
In Button, you can pass a new class into `button_modifiers` (https://github.com/NYPL/nypl-design-system/blob/SFR-694_breadcrumb_react/src/twig/_patterns/01-atoms/buttons/_button.stories.js#L20) that will add in a new class (https://github.com/NYPL/nypl-design-system/blob/SFR-694_breadcrumb_react/src/twig/_patterns/01-atoms/buttons/_button.twig#L20)
However, I don't see anything in breadcrumbs.twig that lets you do anything similar?

helenvholmes replied:
Ohhh, okay, I think I understand better now.

In breadcrumbs, you're able to use that same function whenever there's a class, basically—being able to pass it each of those three parameters. I was kinda lost in the JS here, could but that that works here as well.

helenvholmes reviewed (changes requested):
Let me know if you have any questions. I think this is blocked potentially on the `bem` function in your other PR so I'll get cracking on that as well.

hokei committed:
Moved /react-components to match /twig folder structure

hokei committed:
Merge branch 'development' of https://github.com/NYPL/nypl-design-system into SFR-694_breadcrumb_react

hokei committed:
saving

hokei committed:
saving again

hokei committed:
Add Breadcrumbs

hokei committed:
Merge branch 'development' of https://github.com/NYPL/nypl-design-system into SFR-694_breadcrumb_react

hokei committed:
Removed unused React import

helenvholmes committed:
[694] Update to breadcrumb styles to be on parity with what's in Reno

helenvholmes committed:
[694] Add type scale variables

helenvholmes committed:
[694] Updates from reno in type variables

hokei committed:
saving

helenvholmes committed:
[694] Change breadcrumbs from inline-flex to inline so that our word-break wraps text properly

hokei committed:
Merge branch 'SFR-694_breadcrumb_react' of https://github.com/NYPL/nypl-design-system into SFR-694_breadcrumb_react

helenvholmes committed:
[694] Adds spaceless to remove whitespace that react automatically removes from the browser

helenvholmes committed:
[694] Adds back in margins and padding around the / to breadcrumbs look correct again

hokei committed:
Merge branch 'SFR-694_breadcrumb_react' of https://github.com/NYPL/nypl-design-system into SFR-694_breadcrumb_react

hokei committed:
Added breadcrumb__list class to react

hokei committed:
added changelogs

hokei committed:
Reverted icon.twig changes

hokei commented:
Putting this here: https://github.com/NYPL/nypl-design-system/pull/40/files/0cac9c1b7a98bbf639ec7d012d242d243c7c85a5#diff-b5f9152b3289350c22a6edcf5514f9f8
to make sure we're continuing this conversation.  But also lmk if there's a more sane way to do this via github

helenvholmes commented on src/react-components/CHANGELOG.md:
```diff
@@ -1,5 +1,16 @@
-### v1.0.1
-> Add description
+
+## [1.0.2] - 2019-11-13
```
So a question for our changelogs and Lerna... do we think we'll reset all of the major numbers on these to "0" when we decide we are ready for our first major release?

hokei replied:
I don't know.  NPM won't let us remove and re-publish the package, so if we wanted to do something like this we'd have to pick a different package name.

hokei replied:
Note to self to bump the package.json version numbers

helenvholmes replied:
Renaming the design system is thus a _necessity_ 

`reservoir` here we come

hokei replied:
Maybe we keep calling it a 0 version until we move to the permanent name... ?

hokei replied:
Or maybe we append -prerelease?

hokei replied:
Added "Prerelease" to the changelogs.  Hopefully this is ... ok for what we need for the time being?

helenvholmes replied:
I'm fine with it. I would prefer to have a `0.x.x` but we got what we got!

hokei replied:
If we released under a new package, we could use `0.x.x` - the reason why we weren't doing it was because Lerna treats every pre-1 release as a major release, which ... actually works for us now haha

hokei committed:
Changed package numbers and clarified Prerelease status

helenvholmes commented on src/react-components/src/components/02-molecules/menus/Breadcrumb/Breadcrumb.stories.tsx:
```diff
@@ -0,0 +1,16 @@
+  decorators: [storyFn => <div style={{ backgroundColor: "grey" }}>{storyFn()}</div>],
+};
+
+const shortItems = [{ url: "#", text: "Parent" }, { url: "#", text: "Home" }];
```
Creates longer files, but can we split these up onto multiple lines? I think it makes them a little more legible, especially for `longItems`.

helenvholmes commented on src/react-components/src/components/02-molecules/menus/Breadcrumb/Breadcrumb.tsx:
```diff
@@ -0,0 +1,49 @@
+        role="navigation"
+        aria-label="Breadcrumbs"
+      >
+        <ul className={bem("list", [], breadcrumbs__base_class)}>
```
🌟

helenvholmes commented on src/styles/00-base/02-typography/01-typefaces/_typefaces.scss:
```diff
@@ -1,4 +1,4 @@
-$font-heading: 'HelveticaNeue-Light', 'Arial', sans-serif;
+@import '../../node_modules/system-font-css/system-font.scss';
+
+$font-body: system-ui, "Segoe UI", Tahoma;
```
You shouldn't need to define Segoe and Tahoma here, they're [already built into the font-stack](https://github.com/jonathantneal/system-font-css/blob/gh-pages/system-font.css#L7) that `system-font` declares. More appropriate would just be a `sans-serif` at the end of this imo.

helenvholmes replied:
Man, I'm seeing that it's defined in Reno too... well, let's rip it out here at least.

helenvholmes commented on src/styles/00-base/02-typography/02-type-scale/_type-scale.scss:
```diff
@@ -1 +1,57 @@
-// Type Scale info goes here.
```
Removing this comment is fine (it's no longer descriptive) but I'd rather add some information about how this file _should_ be used. "Type scale information to keep our font-sizes consistent across the system", or something.

helenvholmes commented on src/styles/CHANGELOG.md:
```diff
@@ -1,5 +1,17 @@
-> Add description
+## Prerelease
+
+## [1.0.2] - 2019-11-13
```
Based on our conversation elsewhere, I do think we should be pre-1.0 right now, even though everything is a major release—I mean, everything is a major release! C'est la vie.

hokei replied:
https://semver.org/#spec-item-4 
That's intended (and a thoughtful part of semver!)

helenvholmes replied:
We live, we learn, we laugh, we cry, we sob over our new understanding

hokei replied:
yup, that's how I feel
Thankfully, NPM does let me publish previous versions so we have this lovely https://www.npmjs.com/package/@nypl/design-system-twig

helenvholmes replied:
Does that mean no one will accidentally install our old stuff at the other numbers? This might not be something you know tbf

helenvholmes commented on src/twig/CHANGELOG.md:
```diff
@@ -1,5 +1,17 @@
+All notable changes to this project will be documented in this file.
+
+The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).
+This repo is in Prerelease. When it is released, this project will adhere to [Semantic Versioning](https://semver.org/spec/v2.0.0.html)
```
Can you add a break to match the other Changelog? Also, 👏🏻, good info here

helenvholmes commented on src/twig/CHANGELOG.md:
```diff
@@ -1,5 +1,17 @@
+
+### Changed
+- Breadcrumb uses spaceless to better enforce rendering consistency
+- Breadcrumb uses inline-flex so that word breaks wrap properly
```
I think this line might belong in the `styles` changelog over here. Also we _removed_ the inline flex in favor of just `inline`, I know that happened very quickly during our working session

hokei committed:
Trying to go back to zero version

hokei committed:
Added stuff to the changelog for the test publish

hokei commented:
There is a changelog in root (https://github.com/NYPL/nypl-design-system/blob/development/CHANGELOG.md) I think it makes sense to keep the changelog in the packages so that their consumers have access, but what are your thoughts?

helenvholmes commented on src/react-components/src/utils/bem.ts:
```diff
@@ -0,0 +1,39 @@
+    });
+  }
+
+  let attributes = classes.join(" ");
```
Not for this pull request, but I have some thoughts about how we can combine this and `bem.js`, I think.

helenvholmes commented on src/twig/_patterns/02-molecules/menus/breadcrumbs/breadcrumbs.stories.js:
```diff
@@ -0,0 +1,21 @@
+  // decorators: [storyFn => <div style={{ backgroundColor: "grey" }}>{storyFn()}</div>],
+};
+
+const shortItems = [{ url: "#", text: "Parent" }, { url: "#", text: "Home" }];
```
Same comment as one in the `tsx` stories file.

helenvholmes commented on storybook/.storybook-html/config.js:
```diff
@@ -1,5 +1,5 @@
 import '!style-loader!css-loader!sass-loader!import-glob-loader!design-system-styles/style.scss';
-import { configure } from '@storybook/html';
+import { configure, load, addDecorator } from '@storybook/html';
```
Fingers crossed

helenvholmes reviewed (changes requested):
> There is a changelog in root (https://github.com/NYPL/nypl-design-system/blob/development/CHANGELOG.md) I think it makes sense to keep the changelog in the packages so that their consumers have access, but what are your thoughts?

I think that makes sense. I put it in the root without thinking too deeply clearly!

helenvholmes commented:
Great work, we're almost there. 💫

helenvholmes committed:
[694] Updating the breakpoint for mobile on the breadcrumbs

hokei committed:
Small changes based on CR feedback

hokei committed:
Merging from remote

hokei committed:
Publish

 - @nypl/design-system-react-components@0.0.1
 - @nypl/design-system-styles@0.0.1
 - @nypl/design-system-twig@0.0.1
 - storybook@0.0.1

hokei commented:
The package will be bumped when the release is cut!

helenvholmes commented on src/react-components/src/components/02-molecules/menus/Breadcrumb/Breadcrumb.stories.tsx:
```diff
@@ -0,0 +1,19 @@
+export default {
+  title: "Breadcrumb",
+  component: Breadcrumb,
+  decorators: [storyFn => <div style={{ backgroundColor: "grey" }}>{storyFn()}</div>],
```
Wait, so these are building now????

hokei replied:
it's always built in tsx/React
It's never built in js/Twig (and is still commented out)

helenvholmes committed:
[694] Legibility nits

helenvholmes committed:
[694] Elaborated to include the list-item whitespace bug for when we forget why we did this in two years

helenvholmes committed:
[694] Add back in -body and -heading

helenvholmes commented on src/styles/00-base/02-typography/01-typefaces/_typefaces.scss:
```diff
@@ -1,4 +1,2 @@
-$font-body: 'HelveticaNeue-Light', 'Helvetica Neue', Helvetica, Arial,
-  'Lucida Grande', sans-serif;
-$font-heading: 'HelveticaNeue-Light', 'Arial', sans-serif;
+@import '../../node_modules/system-font-css/system-font.scss';
```
I pushed up a fix that adds those two variables back in—unfortunately we can't delete them even though they reference the same font-stack as they're used somewhere 😬

hokei replied:
It all looks good, yeah, I only realized after I pushed that the font thing wasn't working o.o. 
Anyway I'll push and cut a release tomorrow morning!

helenvholmes reviewed (approved):
_I_ pushed up a few small commits so I think you should take a look at those. If you're good with them, I'm good with all the work you did, this is rad 👍🏻
